### ISSUE TITLE
PySpark DataFrames: Technically immutable, but can be "functionally mutable" if we don't follow best practices.

### ISSUE THREAD

**Issue by JamesDBartlett3** [DataVolumeXYZ/DataVolumeXYZ#49]
**Fun fact:** PySpark DataFrames are [technically immutable](https://web.archive.org/web/20241009132609/https://www.linkedin.com/pulse/common-confusion-spark-dataframes-mutable-lets-clear-up-gaurav-joshi-vq01c/), but they can be "functionally mutable" if we overwrite an existing DataFrame's variable with a modified version of itself.

Even if you overwrite an existing DataFrame's variable (e.g., `df = df.replace('Alice','Annie')`), a new DataFrame with the requested changes will be created in memory, and that new DataFrame will take the place of the existing one in future references to that variable. Meanwhile, the original, unmodified DataFrame is still there, but it's been dereferenced, so you can't access it anymore.

This is unintuitive, and the disconnect between _what's actually happening_ and _what appears to be happening_ can make the code significantly more confusing and difficult to troubleshoot. And that's one of the reasons we have a convention of explicitly declaring a new DataFrame every time we need to make a change to an existing DataFrame -- it forces our code's _apparent behavior_ to more closely resemble its _actual behavior_.

In other words, our DataFrames will actually be immutable no matter what our code appears to be doing, so we might as well write our code in a way that mimics what's actually happening under the hood as closely as possible, and this will make our code easier to debug, enhance, and maintain.
